### ISSUE TITLE
permission issue - mounting files from host

### ISSUE THREAD

**Issue by FelixErnst** [Bioconductor/bioconductor_docker#3]
Hi,

I have problem sorting out a file permission issue when sharing files from the host and the container.

When running with a fixed user and userid

```
docker run -d -it -e USER=$USER -e USERID=$UID -e PASSWORD=bioc -p 8788:8787 -v /home/$USER/GitHub:/home/bioc/github --name biocdevel biocdevel
```

Rstudio reports an error during startup in the browser `Error occurec during transmission`

Without the user specified file permissions are set to 1001:staff, and without setting a userid the file permissions are set to 1001:$usergroup, which is both incompatible with the host user.

I think this is due to the fact that an additional user `bioc` was added, since rocker/rstudio:devel does not exhibit this issue.

Is there any way to fix this during startup of the container?

Best regards
Felix

**Comment by Ziyang-Tan** [comment]
No... It would be really kind if the bioc team could set the username "bioc" to default "rstudio", since a lot of problems may occur. I have the same problem with the issue [https://github.com/rocker-org/rocker-versioned/issues/109](url) and find out that it's the problem of username. I have no idea how to fix it, since changing "USER" would result in the same problem as yours. So I decide to build the image from rocker/rstudio ...

**Comment by nturaga** [comment]
Hi,

When you use the `$USER` and the `$USERID` you are setting brand a new user name and userid. 

Are you seeing the same problem when you start up your container by default? 

```
docker run -d -it -e PASSWORD=bioc -p 8788:8787 -v /home/$USER/GitHub:/home/bioc/github --name biocdevel biocdevel
```

The fact that you run it "interactively" using `-it` should make you log in as `root`. 

Also, you can use the argument `--user rstudio`, if you'd like to use a particular user instead of `bioc`.

```
docker run -d -it --user rstudio -e PASSWORD=bioc -p 8788:8787 -v /home/$USER/GitHub:/home/bioc/github --name biocdevel biocdevel
```

I wasn't able to reproduce your original problem however. Can you point me to how to reproduce it? 

Best,

Nitesh

**Comment by nturaga** [comment]
> No... It would be really kind if the bioc team could set the username "bioc" to default "rstudio", since a lot of problems may occur. I have the same problem with the issue: [rocker-org/rocker-versioned#109](https://github.com/rocker-org/rocker-versioned/issues/109)
> and find out that it's the problem of username. I have no idea how to fix it, since changing "USER" would result in the same problem as yours. So I decide to build the image from rocker/rstudio ...

I suggest you just try and use the `docker run --user rstudio biocondocutor/bioconductor_docker:devel` command, if you insist on using the `rstudio` user. But again, you shouldn't have to set a new $USER or $USERID. I'd like to know the use case of doing this actually.

**Comment by nturaga** [comment]
> Hi,
> 
> I have problem sorting out a file permission issue when sharing files from the host and the container.
> 
> When running with a fixed user and userid
> 
> ```
> docker run -d -it -e USER=$USER -e USERID=$UID -e PASSWORD=bioc -p 8788:8787 -v /home/$USER/GitHub:/home/bioc/github --name biocdevel biocdevel
> ```
> 
> Rstudio reports an error during startup in the browser `Error occurec during transmission`
> 
> Without the user specified file permissions are set to 1001:staff, and without setting a userid the file permissions are set to 1001:$usergroup, which is both incompatible with the host user.
> 
> I think this is due to the fact that an additional user `bioc` was added, since rocker/rstudio:devel does not exhibit this issue.
> 
> Is there any way to fix this during startup of the container?
> 
> Best regards
> Felix

Ok, I got this to work. What username did you try when you tried to sign into Rstudio? It should be whatever user name is being transmitted to RStudio (i.e `$ echo $USER` on your command line).

**Comment by Ziyang-Tan** [comment]
> > No... It would be really kind if the bioc team could set the username "bioc" to default "rstudio", since a lot of problems may occur. I have the same problem with the issue: [rocker-org/rocker-versioned#109](https://github.com/rocker-org/rocker-versioned/issues/109)
> > and find out that it's the problem of username. I have no idea how to fix it, since changing "USER" would result in the same problem as yours. So I decide to build the image from rocker/rstudio ...
> 
> I suggest you just try and use the `docker run --user rstudio biocondocutor/bioconductor_docker:devel` command, if you insist on using the `rstudio` user. But again, you shouldn't have to set a new $USER or $USERID. I'd like to know the use case of doing this actually.

Hi, 
I have tried `docker run \
        --rm \
        --user rstudio \
        -v /Some/folder:/home/rstudio/somefolder \
        -e PASSWORD=bioc \
        -e ROOT=TRUE \
        -p 8787:8787 \
        bioconductor/bioconductor_docker:devel`
and get an error : "s6-mkdir: warning: unable to mkdir /var/run/s6: Permission denied". and adding "sudo" to the command doesn't help.

**Comment by nturaga** [comment]
> > > No... It would be really kind if the bioc team could set the username "bioc" to default "rstudio", since a lot of problems may occur. I have the same problem with the issue: [rocker-org/rocker-versioned#109](https://github.com/rocker-org/rocker-versioned/issues/109)
> > > and find out that it's the problem of username. I have no idea how to fix it, since changing "USER" would result in the same problem as yours. So I decide to build the image from rocker/rstudio ...
> > 
> > 
> > I suggest you just try and use the `docker run --user rstudio biocondocutor/bioconductor_docker:devel` command, if you insist on using the `rstudio` user. But again, you shouldn't have to set a new $USER or $USERID. I'd like to know the use case of doing this actually.
> 
> Hi,
> I have tried `docker run --rm --user rstudio -v /Some/folder:/home/rstudio/somefolder -e PASSWORD=bioc -e ROOT=TRUE -p 8787:8787 bioconductor/bioconductor_docker:devel`
> and get an error : "s6-mkdir: warning: unable to mkdir /var/run/s6: Permission denied". and adding "sudo" to the command doesn't help.

 Yes, i get this too. It seems like you can only access rstudio user interactively 

```
docker run -it --user=rstudio bioconductor/bioconductor_docker:devel bash
```

But i'm still not sure why you need the `rstudio` user?

**Comment by nturaga** [comment]
@Ziyang-Tan 

Remember that the username of the Rstudio being started up is `bioc`, password is whatever you set. I think that was the `Error occurec during transmission` issue.

**Comment by FelixErnst** [comment]
> I have tried `docker run --rm --user rstudio -v /Some/folder:/home/rstudio/somefolder -e PASSWORD=bioc -e ROOT=TRUE -p 8787:8787 bioconductor/bioconductor_docker:devel`
> and get an error : "s6-mkdir: warning: unable to mkdir /var/run/s6: Permission denied". and adding "sudo" to the command doesn't help.

Same here. I omited this behavior from my initial post, since that was the obvious solution, which does not work.

```
docker run --user rstudio biocondocutor/bioconductor_docker:devel
```

works of course, but then I don't have any mounted folders. 

When I start the container with

```
docker run -d -it -e PASSWORD=bioc -p 8788:8787 -v /home/$USER/GitHub:/home/bioc/github --name biocdevel biocdevel
```

I can login as rstudio, but not with the password `bioc`, rather the default pw `rstudio`.

**Comment by FelixErnst** [comment]
> > > > No... It would be really kind if the bioc team could set the username "bioc" to default "rstudio", since a lot of problems may occur. I have the same problem with the issue: [rocker-org/rocker-versioned#109](https://github.com/rocker-org/rocker-versioned/issues/109)
> > > > and find out that it's the problem of username. I have no idea how to fix it, since changing "USER" would result in the same problem as yours. So I decide to build the image from rocker/rstudio ...
> > > 
> > > 
> > > I suggest you just try and use the `docker run --user rstudio biocondocutor/bioconductor_docker:devel` command, if you insist on using the `rstudio` user. But again, you shouldn't have to set a new $USER or $USERID. I'd like to know the use case of doing this actually.
> > 
> > 
> > Hi,
> > I have tried `docker run --rm --user rstudio -v /Some/folder:/home/rstudio/somefolder -e PASSWORD=bioc -e ROOT=TRUE -p 8787:8787 bioconductor/bioconductor_docker:devel`
> > and get an error : "s6-mkdir: warning: unable to mkdir /var/run/s6: Permission denied". and adding "sudo" to the command doesn't help.
> 
> Yes, i get this too. It seems like you can only access rstudio user interactively
> 
> ```
> docker run -it --user=rstudio bioconductor/bioconductor_docker:devel bash
> ```
> 
> But i'm still not sure why you need the `rstudio` user?

Because that user has the userid 1000:1000, which by accident matches the userid of my user I start the container from. And only in this situation, I end up without permission issues, when mounting project files from the host.

I know, this setup is probably weird and a bit unhealthy, but it worked previously and my initial question stands (which was not spelled out, I admit): How do I avoid a permission conflict using a mounted folder from the host?

Again, previously this question was solved by accident, but now it has come up.

**Comment by nturaga** [comment]
> > I have tried `docker run --rm --user rstudio -v /Some/folder:/home/rstudio/somefolder -e PASSWORD=bioc -e ROOT=TRUE -p 8787:8787 bioconductor/bioconductor_docker:devel`
> > and get an error : "s6-mkdir: warning: unable to mkdir /var/run/s6: Permission denied". and adding "sudo" to the command doesn't help.
> 
> Same here. I omited this behavior from my initial post, since that was the obvious solution, which does not work.
> 
> ```
> docker run --user rstudio biocondocutor/bioconductor_docker:devel
> ```
> 
> works of course, but then I don't have any mounted folders.
> 
> When I start the container with
> 
> ```
> docker run -d -it -e PASSWORD=bioc -p 8788:8787 -v /home/$USER/GitHub:/home/bioc/github --name biocdevel biocdevel
> ```
> 
> I can login as rstudio, but not with the password `bioc`, rather the default pw `rstudio`.

You'll have to walk me through this again i'm afraid, i'm just at the point of trying to reproduce what you see here. I'm trying,

```
$ docker run -d -it -e PASSWORD=bioc -p 8788:8787 -v /Users/ni41435_lca/Documents:/home/bioc/github bioconductor/bioconductor_docker:devel

3f6f41e43be228f999045b34dde5dd2672953e6ad378d0c3791267da54a5578e
```
This starts a docker process in daemon mode. And i'm able to login with username `bioc` and password `bioc`.

![Screen Shot 2020-01-30 at 11 58 09 AM](https://user-images.githubusercontent.com/2746443/73471771-e7e54700-4357-11ea-9474-26f7b2c7fb17.png)
![Screen Shot 2020-01-30 at 11 58 30 AM](https://user-images.githubusercontent.com/2746443/73471782-eb78ce00-4357-11ea-986c-3f6f6da7ad9d.png)

**Comment by LiNk-NY** [comment]
Hi Felix, @FelixErnst 
The Docker containers are not meant to mount an alternative user's
folders / packages. 
You'd have to install all of the packages that you use via the Docker
container to avoid this issue. _This is the expected use case._ I have
tried to rebuild the Docker image in the past so that the IDs match but
you still encounter issues with shared libraries...

**Comment by FelixErnst** [comment]
Hi Nitesh, @nturaga 

add a file under the github folder, go back to the host and check the file's permission. The will probably come back as `1001:$(theUsernameOnTheHost)`.

With this, you cannot copy, you cannot work with git, which is/was part of my personal use case.

**Comment by FelixErnst** [comment]
> Hi Felix, @FelixErnst
> The Docker containers are not meant to mount an alternative user's
> folders / packages.
> You'd have to install all of the packages that you use via the Docker
> container to avoid this issue. _This is the expected use case._ I have
> tried to rebuild the Docker image in the past so that the IDs match but
> you still encounter issues with shared libraries...

Well, how do I get my GitHub version to R-devel on a Windows system, running VBox, running docker?

Previously that worked find.

**Comment by LiNk-NY** [comment]
@FelixErnst 
You can create a docker home folder `DOCKER_HOME="$HOME/dockerhome"`, add your config files to the folder and mount it with `-v $DOCKER_HOME:/home/bioc`. Would that work?

**Comment by Ziyang-Tan** [comment]
@nturaga 
for me, `docker run -it --user=rstudio bioconductor/bioconductor_docker:devel` turns out the same error "s6-mkdir: warning: unable to mkdir /var/run/s6: Permission denied".

The reason I would like to use default username is that, when I use bioc to login rstudio with a mounted volume with "-v" command, I cannot write from R studio to the mounted folder. i.e. I cannot see a newly created R-script in the host's file system. But I can see the host's folder and file in R studio.
This is the same problem with another issue: https://github.com/rocker-org/rocker-versioned/issues/109 . While in this issue, the solution seems to be changing the username to default. 

@LiNk-NY
you're right that docker container is made to separate packages, i.e. the environment. However, I think a shared folder is convenient for data analysis.

Actually, my problem is done by build my environment from rocker/rstudio de novo, but I'm glad to find out if it can be solved in bioconductor/bioconductor_docker.

**Comment by LiNk-NY** [comment]
@Ziyang-Tan 
See this for more information on that error: https://github.com/lablup/backend.ai-agent/issues/103

I agree that it would be more convenient to load / attach a volume of packages that are ready to use.
The best bet would be to have a volume of binaries that can be easily installed to the container.

**Comment by FelixErnst** [comment]
Sure, but than I just have a different folder within which I have to change the file permissions to be able to works with the files on the host.

![image](https://user-images.githubusercontent.com/17824332/73473803-ab344280-438d-11ea-9f88-c5bd62092353.png)

So, every which way I turn this, it ends up as a permission problem. The only feasible and possibly long term solution is to be able to use the command, which I used in my initial post:

```
docker run -d -it -e USER=$USER -e USERID=$UID -e PASSWORD=bioc -p 8788:8787 -v /home/$USER/GitHub:/home/bioc/github --name biocdevel biocdevel
```

With this solution I would be able to use whatever user id I choose and therefore it would be compatible with every user on any machine. The solution described above, was also mentioned in the documentation for the deprecated bioconductor docker images (although I cannot find a reference, and this might be my imagination) and the underlying rocker versions (See https://github.com/rocker-org/rocker-versioned/blob/master/rstudio/README.md#additional-configuration-options). My guess would be, that it worked with the deprecated docker images as well, but I have never tested this.

**Comment by FelixErnst** [comment]
Ok now I didn't detach the docker start up and the following message appears:

```
felix@felix-VirtualBox:~$ docker run -it -e USERID=1000 -e PASSWORD=bioc -p 8789:8787 -v /home/felix/dockerhome:/home/bioc --name biocdevel2 biocdevel
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] add: executing... 
Nothing additional to add
[cont-init.d] add: exited 0.
[cont-init.d] userconf: executing... 
usermod: user 'bioc' already exists
USER is now bioc
[cont-init.d] userconf: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
```

The bioc user is somehow hard hardcoded into the image and cannot be changed.

**Comment by LiNk-NY** [comment]
@FelixErnst 
Yes, we added the `bioc` user to `bioconductor_docker`. 

I meant that you can just read your git config / certificate  files from the home
folder and download packages using `git` as the intended use case. 

It's hard to follow without a minimally reproducible example.
Would you like to work on one? 
Including file creation on host user and how you are accessing the files afterward? 
and without the use of `--name biocdevel biocdevel`.

**Comment by FelixErnst** [comment]
I will give it a try.

1. Create a folder `/home/$USER/dockerhome`
2. `docker run -it -e USER=$USER -e PASSWORD=bioc -p 8788:8787 -v /home/$USER/dockerhome:/home` 
3. Go to your browser and access RStudio.  It will generate a user initialization error.
4. Check you console output. Something like this should appear:
```
30 Jan 2020 17:58:29 [rsession-bioc] ERROR system error 13 (Permission denied) [path=/home/bioc/.rstudio, target-dir=]; OCCURRED AT: rstudio::core::Error rstudio::core::FilePath::createDirectory(const string&) const /var/lib/jenkins/workspace/IDE/open-source-pipeline/v1.2-patch/src/cpp/core/FilePath.cpp:846; LOGGED FROM: rstudio::core::FilePath rstudio::core::system::userSettingsPath(const rstudio::core::FilePath&, const string&, bool) /var/lib/jenkins/workspace/IDE/open-source-pipeline/v1.2-patch/src/cpp/core/system/PosixSystem.cpp:491
```
5. Check content and file permissios of  `/home/$USER/dockerhome`. There should be a folder with correct access rights set (e.g. `felix:felix`)
6. delete the `dockerhome` folder
7. repeat with the following combinations of values for `-e USERID=` instead of `-e USER=`
    - `-e USERID=1000`
    - `-e USERID=1001`

The last options works since this is the `bioc` user. According to the startup message posted above this is hardcoded into the `bioconductor_docker` image.

```
[cont-init.d] userconf: executing... 
usermod: user 'bioc' already exists
USER is now bioc
```

This presents a problem in usability as described above regarding file mounting and permissions. It also deviates from the documenation, which I assume is also inherited from the rocker image (See https://github.com/rocker-org/rocker-versioned/blob/master/rstudio/README.md#additional-configuration-options). Therefore, if I may be so bold, two options exist:

1. amend the documentation and stated that this docker image is hardcoded to as user with the user id `1001` with all its implications regarding file mounting.
2. rework the image so that bioc user exists, but respects the values for `USER`, `USERID` and so on set in the environment.


<details>
<pre>
felix@felix-VirtualBox:$ docker version
Client: Docker Engine - Community
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        633a0ea838
 Built:             Wed Nov 13 07:29:52 2019
 OS/Arch:           linux/amd64
 Experimental:      false
Server: Docker Engine - Community
 Engine:
  Version:          19.03.5
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.12
  Git commit:       633a0ea838
  Built:            Wed Nov 13 07:28:22 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
felix@felix-VirtualBox:$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.3 LTS
Release:	18.04
Codename:	bionic
</pre>
</details>

**Comment by snystrom** [comment]
Just to chime in here, I am having this same issue on systems where my UID is not 1001. 

I can fix the issue if I rebuild the bioconductor_docker image after removing the steps from the Dockerfile to create and set the user to `bioc`. I really don't know why it's necessary to make this new user except to call it `bioc` for vanity purposes. The base `rocker` images handle UID issues themselves, and they already do the setup of `user-settings/`. Is there additional benefit to the `bioc` user that I am missing?

**Comment by nturaga** [comment]
Hi, @FelixErnst @Ziyang-Tan @snystrom 

The latest fix removes the `bioc` user. https://github.com/Bioconductor/bioconductor_docker/commit/3d4b0212129f9de1522152d3e562513b08384506

Please update your image (pull) again and this should be fixed.

**Comment by FelixErnst** [comment]
Thank you. I am sorry for the convoluted way of getting a good description and reproducible example. Should have done that from the beginning.

**Comment by snystrom** [comment]
Thanks for the prompt response!

**Comment by nturaga** [comment]
If you are satisfied with the new update and image, feel free to close the
issue.
